Replace the 32-bit scalar field with a fiat-generated implementation

gtank opened the pull request:
As the title. WIP.

FiloSottile commented on scalar_test.go:
```diff
@@ -217,7 +227,8 @@ type notZeroScalar Scalar
 	var s Scalar
-	for s == scZero {
+	var isNonZero uint64
+	for fiat_sc255_nonzero(&isNonZero, (*[4]uint64)(&s.s)); isNonZero == 0; {
```
```suggestion
	for isNonZero == 0 {
		s = Scalar{}.Generate(rand, size).Interface().(Scalar)
		fiat_sc255_nonzero(&isNonZero, (*[4]uint64)(&s.s))
	}
```

gtank replied:
d'oh!

gtank commented:
Fixed some silly mistakes and updated to the latest generated fiat code, which has improved a lot since I first did this over a year (!) ago. In particular, it's introduced some types to differentiate between Montgomery / non-Montgomery field elements and seems to be faster across the board.

The major remaining issue is that we use a double-width reducing decode function that the festive code calls `scReduce` and fiat doesn't generate. In the future I'd like to replace that with 64-bit code too, but it works for now.

These benchmarks are new and pretty narrow, but suggest this is well worth doing!

🎄 vs year-old fiat field:
```
$ benchstat bench-noboost-baseline bench-noboost-old_fiat 
name                            old time/op  new time/op  delta
ScalarAddition-16                105ns ± 0%     7ns ± 0%  -93.13%  (p=0.000 n=9+10)
ScalarMultiplication-16          106ns ± 0%    44ns ± 1%  -58.96%  (p=0.000 n=7+10)
ScalarInversion-16              30.7µs ± 0%  12.7µs ± 1%  -58.60%  (p=0.000 n=9+10)
```

🎄 vs current fiat field:
```
$ benchstat bench-noboost-baseline bench-noboost-new_fiat 
name                            old time/op  new time/op  delta
ScalarAddition-16                105ns ± 0%     6ns ± 0%  -94.64%  (p=0.000 n=9+10)
ScalarMultiplication-16          106ns ± 0%    40ns ± 0%  -62.61%  (p=0.000 n=7+9)
ScalarInversion-16              30.7µs ± 0%  11.7µs ± 0%  -61.81%  (p=0.000 n=9+10)
```

Year-old fiat vs current fiat:
```
$ benchstat bench-noboost-old_fiat bench-noboost-new_fiat 
name                            old time/op  new time/op  delta
ScalarAddition-16               7.22ns ± 0%  5.63ns ± 0%  -21.93%  (p=0.000 n=10+10)
ScalarMultiplication-16         43.5ns ± 1%  39.7ns ± 0%   -8.90%  (p=0.000 n=10+9)
ScalarInversion-16              12.7µs ± 1%  11.7µs ± 0%   -7.74%  (p=0.000 n=10+10)
```

gtank commented:
Oh, noting here for later: it would be nice if `fiat_sc255_to_bytes` took a `fiat_sc255_non_montgomery_domain_field_element` as suggested by its documentation, rather than a raw `*[4]uint64`. It would save an ugly cast in the following usage

```
func (s *Scalar) bytes(out *[32]byte) []byte {
	var limbs fiat_sc255_non_montgomery_domain_field_element
	fiat_sc255_from_montgomery(&limbs, &s.s)
	fiat_sc255_to_bytes(out, (*[4]uint64)(&limbs))
	return out[:]
}
```

FiloSottile commented:
Replaced `scReduce` with a series of operations reducing the value in limbs, following the identity `x = a + b * 2^168 + c * 2^336 mod l`. Updated the fiat-crypto code to skip some type-casting that generates worse code, probably improves performance.

FiloSottile commented:
Excluding autogenerated code, 262 insertions(+), 866 deletions(-) 🎉✨💥
